Add anchor support for dynamic blocks

Soean opened the pull request:
## What?

- Adds [anchor](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#anchor) support to dynamic rendered blocks.
- Enables support for core blocks

Fixes #29401

## Testing Instructions

You need to add `'id'` to `$attributes_to_merge` in `get_block_wrapper_attributes()` from the `WP_Block_Supports` class, see https://github.com/WordPress/wordpress-develop/blob/6.0.2/src/wp-includes/class-wp-block-supports.php#L178
 ```php
$attributes_to_merge = array( 'style', 'class' );
```
Change to:
 ```php
$attributes_to_merge = array( 'style', 'class', 'id' );
```

Soean committed:
Add anchor support

Soean committed:
Add anchor support

michalczaplinski commented:
I'm very eager to get this merged! Nice work @Soean! 👏 

> You need to add 'id' to $attributes_to_merge in get_block_wrapper_attributes() from the WP_Block_Supports class, see https://github.com/WordPress/wordpress-develop/blob/6.0.2/src/wp-includes/class-wp-block-supports.php#L178

Could you create a ticket for this in [WordPress Core Trac](https://core.trac.wordpress.org/) so that a relevant change can be made there? 

There's also a small merge conflict in the docs I see 🙂

Soean committed:
Merge trunk

Soean committed:
Merge trunk

Soean commented:
@michalczaplinski I created a trac ticket and PR:

- https://core.trac.wordpress.org/ticket/56852
- https://github.com/WordPress/wordpress-develop/pull/3497

I also solved the merge conflict.

ndiego commented:
Just tested this PR using a Social Link block, and it works great. For others testing, note that you need to manually apply the patch in https://github.com/WordPress/wordpress-develop/pull/3497.

**Editor:**
![image](https://user-images.githubusercontent.com/4832319/215613535-e65431c3-7440-4c09-b16c-5afb815a86de.png)

**Front end:**
![image](https://user-images.githubusercontent.com/4832319/215613881-91f0e49c-7e51-409a-8335-a954b55224d8.png)

carolinan commented:
Would it be possible to make the anchor block support enabled by default for all blocks?

carolinan committed:
Merge branch 'trunk' into add-anchor-support

Mamaduka commented:
@ndiego, do you see any errors in the DevTools console when testing this branch, with or without the core patch?

I think failing E2E tests are related to the changes in this PR.

carolinan committed:
Remove duplicate block support after merge conflict

ndiego commented:
@carolinan I am getting fatal errors now with the recent updates. Per @Mamaduka it looks like this PR needs to be updated due to https://github.com/WordPress/gutenberg/pull/47195.

carolinan committed:
Use wp_should_skip_block_supports_serialization

carolinan commented:
After updating to `wp_should_skip_block_supports_serialization`
I am still seeing this console warning in the editor:
` traverse.js:26 Error while traversing the CSS: Error: undefined:1:10: missing '{'`

**Update: Ignore that ^**, it was theme related.

ndiego commented:
I just retested and everything is looking good on my end.

Soean commented:
Thanks for testing, what do we need to get it merged?

Mamaduka reviewed (approved):
The changes here look good to me. Thank you, @Soean!

While there are many file updates, only the [`anchor.php`](https://github.com/WordPress/gutenberg/pull/44771/files#diff-ead267de3e548aaf5144199e67b58d49e1eedc838ce77c35954930b9edd4613a) contains significant changes.

@Soean, can you update https://github.com/WordPress/wordpress-develop/pull/3497 and sync changes for the `anchor.php` after the merge?

Here's the tracking issue for PHP backports: https://github.com/WordPress/gutenberg/issues/47187.

Soean commented:
Thanks @Mamaduka.  I synced the changes to the PR https://github.com/WordPress/wordpress-develop/pull/3497, updated the trac ticket and added it to the tracking issue.

Mamaduka commented:
Thank you, @Soean!

fabiankaegy commented:
@Soean I added the `Needs Dev Note` label to this PR. Do you have time to write something up for it in the next few weeks (Deadline is March 3rd :) )

If not I can help with writing something up. :)

t-hamano commented:
I fear that this change may affect the backward compatibility of some blocks with the anchor. Please see #48232.

t-hamano commented on lib/block-supports/anchor.php:
```diff
@@ -0,0 +1,67 @@
+ *
+ * @param WP_Block_Type $block_type Block Type.
+ */
+function gutenberg_register_anchor_support( $block_type ) {
+	$has_anchor_support = _wp_array_get( $block_type->supports, array( 'anchor' ), true );
+	if ( ! $has_anchor_support ) {
+		return;
+	}
+
+	if ( ! $block_type->attributes ) {
+		$block_type->attributes = array();
+	}
+
+	if ( ! array_key_exists( 'anchor', $block_type->attributes ) ) {
+		$block_type->attributes['anchor'] = array(
+			'type' => 'string',
+		);
+	}
+}
```
This process would also apply to static blocks, which would add an extra comment as attributes.

adamwoodnz commented:
> I fear that this change may affect the backward compatibility of some blocks with the anchor. Please see #48232.

We've been building pages for wordpress.org and seeing blocks with existing anchors becoming unrecoverable. Typically  Group blocks.

bph commented:
@fabiankaegy and @Soean Are either of you [working the dev note](https://github.com/WordPress/gutenberg/pull/44771#issuecomment-1418795101) for this PR? 

- Does it warrant a stand-alone post? If yes, please post a draft to the Make Core Blog. 
- If you only would write a paragraph or two, comment on this PR, and I'll add it to the Misc Editor Note. 

If you could do write this by Thursday (March 2 EOB), that would be helpful. The release squad for 6.2 documentation, will start working on the Field guide for 6.2 on Friday.

t-hamano commented:
@bph 
In this PR, we found a backward compatibility issue where blocks with anchors were broken, as reported in #47771. We are still discussing #48438 to fix this issue and how it should be addressed.

t-hamano commented:
@bph
This change will be reverted in #48592 due to backward compatibility issues as reported in #48232. [Here](https://wordpress.slack.com/archives/C02QB2JS7/p1677581267400419) is the discussion on Slack. Therefore, I think Dev Note isn't necessary.
